### PR TITLE
Bug 1690780 - Coverage support

### PR DESCRIPTION
**Pull request by mdboom**
Just a playground to see if I can get the pieces working together.

Depends on https://github.com/mozilla/glean_parser/pull/272 for the glean_parser side of things.

Example coverage uploaded to codecov.io here: https://codecov.io/gh/mozilla/glean/src/ed8704ac50bcdfcb44ddd3a1fd6004d0eef57675/glean-core/metrics.yaml

Unfortunately, this requires forking the codecov.io uploader, since it explicitly filters out coverage reporting on `yaml` files.  (In fairness "coverage" on yaml files is unusual, but seems like it shouldn't be out-and-out disallowed).  Anyway, we might have to address that upstream in order for codecov.io to be a viable solution for our users of this feature.

Other than that, the implementation here turned out much simpler than in the proposal (reminder to go back and update the proposal with where we ended up).  By using only a environment variable to turn the feature on (thanks @brizental for the suggestion) and just writing to a file as we go (thanks @chutten for the suggestion), there's no need to hook into the unit testing framework, and thus there's no need for an FFI or language bindings, which makes this all over quite a small change.

### CODE REVIEW

File: .circleci/config.yml
Comment (mdboom) [inline .circleci/config.yml:73]: This just installs a branch of glean_parser.  No longer necessary once that branch is merged and we make a release.

File: .circleci/config.yml
Comment (mdboom) [inline .circleci/config.yml:63]: Have to use `realpath` here since `cargo test` changes the working directory.  This makes sure there is a single file (even though there are multiple crates tested) and it ends up where we expect it to.

File: .circleci/config.yml
Comment (mdboom) [inline .circleci/config.yml:75]: Had to fork codecov.io's upload script so it wouldn't ignore `yaml` files...

**Comment by badboy** [comment]
initial commit that strips out yaml files in the uploader is here: https://github.com/codecov/codecov-bash/commit/aa5f5839776e02769eac09aa02f3360799da6eae

File: glean-core/src/coverage.rs
Comment (badboy) [inline glean-core/src/coverage.rs:3]: Can we add some module docs here (`//!`) of what this does. Calling out the environment variable and the format in which the file is written ("one metric id per line")

File: glean-core/src/storage/mod.rs
Comment (badboy) [inline glean-core/src/storage/mod.rs:146]: The proposal said `snapshot_metric` can be called from one non-test place. Do we need to handle this here?

**Review by badboy — COMMENTED** [review]
This is ... too simple. :D 

I guess the difference to the proposal is that we don't need to keep any state beyond the `GLEAN_TEST_COVERAGE` file part?
When we want to integrate this into test harnesses, we could just say "set this before you run your tests" or we can build out some helpers (such as our GleanTestRule) that sets that environment variable, right?)

**Comment by mdboom** [comment]
> When we want to integrate this into test harnesses, we could just say "set this before you run your tests" or we can build out some helpers (such as our GleanTestRule) that sets that environment variable, right?)

Yeah, I think depending on the details, they could just set the environment variable before running the test harness (as here).  I haven't tested that with Gradle/JUnit yet because my Android build environment seems to have broken...

**Comment by mdboom** [comment]
The upstream changes in glean_parser and codecov.io bash uploader have landed, so this is ready for a final review.

File: CHANGELOG.md
Comment (badboy) [inline CHANGELOG.md:10]:
```suggestion
  * The Glean SDK can now produce testing coverage reports for your metrics ([#1482](https://github.com/mozilla/glean/pull/1482/files)).
```

Comment (badboy) [reply CHANGELOG.md:10]: Should we add a link to the book here?

File: docs/dev/code_coverage.md
Comment (badboy) [inline docs/dev/code_coverage.md:8]: Should this link to the user documentation?

File: docs/user/user/testing-metrics.md
Comment (badboy) [inline docs/user/user/testing-metrics.md:291]: Should this call out that (right now) we only support the codecov format?

**Review by badboy — APPROVED** [review]
Good from my side, couple of questions, mostly around documentation.

**Pull request merged**
